src/constraintmatchnode.jl

```diff
@@ -0,0 +1,15 @@
+"""
+Tree structure to which rulenode trees can be matched.
+"""
+abstract type AbstractConstraintMatchNode end
+
+struct ConstraintMatchNode <: AbstractConstraintMatchNode
+    rule_ind::Int
+    children::Vector{AbstractConstraintMatchNode}
+end
+
+ConstraintMatchNode(rule_ind::Int) = ConstraintMatchNode(rule_ind, [])
+
+struct ConstraintMatchVar <: AbstractConstraintMatchNode
```

Comment:
Variable means "it can match any ruleNode as long as it has a correct root type"?

Reply:
Not necessarily, the symbol is an actual variable name and doesn't have to do anything with the type. This allows you to forbid trees such as `a+a`.
You can also use it as a 'match anything', e.g. `a+b` forbids every addition.

Reply:
alright, that makes perfect sense! Maybe include that as a comment?